File: python-for-cote/5/3/sol.py
import sys
n ,m = map(int, input().split())
graph = []
for i in range(n):
    graph.append(list(map(int, input())))####### 배열 입력

print("graph", graph)
### 풀이 가이드
# 1. 특정 지점의 주변 상, 하, 좌, 우를 살펴본 뒤에 주변 지점 중에서 값이 '0' 이면서 아직
# 1-1방문하지 않은 지점이 있다면 해당 지점을 방문한다

# 2. 방문한 지점에서 다시 상하좌우를 살펴보면서 방문을 다시 진행하면 연결된 모든 지점을 방문하지 않은 지점의 수를 센다


def dfs(x, y):
    if x < 0 or x >= n or y < 0 or y >= m:
        return False 
            # wrong
    if graph[x][y] == 0: # 방문 안했으면 
        graph[x][y] = 1  # 방문처리 
        dfs(x-1, y) # 하
        dfs(x+1, y) # 상
        dfs(x, y-1) # 좌
        dfs(x, y+1) # 우
        return True 
    return False

result = 0
for i in range(n):
    for j in range(m):
        if dfs(i, j) == True:
            result += 1

print(result)

File: python-for-cote/5/3/sol.cpp
#include <bits/stdc++.h>

using namespace std;

int n, m;
int graph[100][100];

bool dfs(int x, int y){
    // 주어진 범위를 벗어나는 경우에는 즉시 종료
    if (x < 0 || x >= n || y < 0 || y >=m){
        return false;
    }

    if(graph[x][y] == 0){
        // 해당 노드 방문 처리 
        graph[x][y] = 1;
        dfs(x-1, y);
        dfs(x, y-1);
        dfs(x+1, y);
        dfs(x, y+1);
        return true;
    }
    return false;
}

int main(){
    // N, M을 공백 기준으로 구분하여 입력 받기
    cin >> n >> m;
    // 2차원 리스트의 맵 정보 입력 받기
    for(int i = 0; i < n; i++){
        for (int j = 0; j < m; j++){
            scanf("%1d", &graph[i][j]);
        }
    }
    // 모든 노드에 대해서 음료수 채우기

    int result = 0;
    for (int i = 0; i < n; i++){
        for (int j = 0; j < m; j++){
            // 현재 위치에서 DFS 수행 
            if (dfs(i, j)){
                result += 1;
            }
        }
    }

    cout << result << '\n';
}